contributors/design-proposals/daemonset-update.md

```diff
@@ -0,0 +1,196 @@
+	// daemon pod and are supposed to run the daemon pod.
+	CurrentNumberScheduled int32
+
+	// NumberMisscheduled is the number of nodes that are running the daemon pod, but are
+	// not supposed to run the daemon pod.
+	NumberMisscheduled int32
+
+	// DesiredNumberScheduled is the total number of nodes that should be running the daemon
+	// pod (including nodes correctly running the daemon pod).
+	DesiredNumberScheduled int32
+
+	// NumberReady is the number of nodes that should be running the daemon pod and have one
+	// or more of the daemon pod running and ready.
+	NumberReady int32
+
+	// ObservedGeneration is the most recent generation observed by the daemon set controller.
+	ObservedGeneration int64
+
+	// UpdatedNumberScheduled is the total number of nodes that are running updated
+	// daemon pod
+	UpdatedNumberScheduled int32 `json:"updatedNumberScheduled"`
+
+}
+```
+
+### Controller 
+
+#### DaemonSet Controller 
+
+The DaemonSet Controller will make DaemonSet updates happen. It will watch
+DaemonSets in etcd. 
+
+For each pending DaemonSet updates, it will:
+
+1. Find all pods whose label is matched by `DaemonSetSpec.Selector`. 
+1. Find all nodes that should run these pods created by this DaemonSet.
+1. Find an existing PodTemplate whose `Template` is the same as
+   `DaemonSetSpec.Template`
+   - If not found, create a new PodTemplate from `DaemonSetSpec.Template`
+   - The name will be `<DaemonSet-Name>-<Hash-of-pod-template>`
+   - Add "pod-template-hash" labels to `PodTemplate.Metadata.Labels`, value be
+     the hash of `PodTemplate.Spec`
+1. Create daemon pods on nodes when they should have those pods running but not
+   yet, and add the same "pod-template-hash" labels to their `Metadata.Labels`.
+   Otherwise, delete running daemon pods that shouldn't be running on nodes. 
+1. Check `DaemonSetUpdateStrategy`:
+   - If `OnDelete`: do nothing
+   - If `RollingUpdate`:
+     - From all nodes that should run daemon pods, check the daemon pod's
+       "pod-template-hash" label, and kill it if it doesn't equal to the hash of
+       `DaemonSetSpec.Template.Spec` and if it won't violate `MaxUnavailable`.
+     - Go back to step 1 
+1. Cleanup 
+
+If DaemonSet Controller crashes during an update, it can still recover. 
+
+
+## Updating DaemonSets mid-way
+
+Users can update an updated DaemonSet before its rollout completes.
+In this case, the existing daemon pods will not continue rolling out and the new
+one will begin rolling out.
+
+
+## Deleting DaemonSets
+
+Deleting a DaemonSet (with cascading) will delete all its pods and podtemplates. 
+
+
+## DaemonSet Strategies
+
+DaemonSetStrategy specifies how the new daemon pods should replace existing ones.
+To begin with, we will support 2 types:
+
+* On delete: Do nothing, until existing daemon pods are killed (for backward
+  compatibility).
+* Rolling update: We gradually kill existing ones while creating the new one.
+
+
```

Comment:
Should the DS controller cordon the node even if a drain is not required?  This will ensure that a non-daemon pod is not scheduled into the free space created by deleting the old pod, before the new pod can start.